include/gridtools/stencil_composition/backend_cuda/entry_point.hpp

```diff
@@ -0,0 +1,215 @@
+                    using info_t = decltype(info);
+                    return make_tmp_storage<typename info_t::data_t>(typename info_t::num_colors_t(),
+                        i_block_size_t(),
+                        j_block_size_t(),
+                        extent_t(),
+                        n_blocks_i,
+                        n_blocks_j,
+                        k_size,
+                        allocator);
+                },
+                hymap::from_keys_values<meta::transform<stage_matrix::get_plh, plh_map_t>, plh_map_t>());
+        }
+
+        template <class DataStoreMap>
+        auto block(DataStoreMap data_stores) {
+            return tuple_util::transform(
+                [](auto &&src) {
+                    return sid::block(std::forward<decltype(src)>(src),
+                        hymap::keys<dim::i, dim::j>::values<i_block_size_t, j_block_size_t>());
+                },
+                std::move(data_stores));
+        }
+
+        template <class Is, class... Funs>
+        struct multi_kernel;
+
+        template <class... Funs, size_t... Is>
+        struct multi_kernel<std::index_sequence<Is...>, Funs...> {
+            tuple<Funs...> m_funs;
+
+            template <class Validator>
+            GT_FUNCTION_DEVICE void operator()(int_t i_block, int_t j_block, Validator const &validator) const {
+                (void)(int[]){
+                    (__threadfence_block(), tuple_util::device::get<Is>(m_funs)(i_block, j_block, validator), 0)...};
+            }
+        };
+
+        template <class Fun>
+        Fun make_multi_kernel(tuple<Fun> tup) {
+            return tuple_util::get<0>(std::move(tup));
+        }
+
+        template <class... Funs>
+        multi_kernel<std::index_sequence_for<Funs...>, Funs...> make_multi_kernel(tuple<Funs...> tup) {
+            return {std::move(tup)};
+        }
+
+        template <int_t BlockSize>
+        std::enable_if_t<BlockSize == 0, int_t> blocks_required_z(int_t) {
+            return 1;
+        }
+
+        template <int_t BlockSize>
+        std::enable_if_t<BlockSize != 0, int_t> blocks_required_z(int_t nz) {
+            return (nz + BlockSize - 1) / BlockSize;
+        }
+
+        template <class MaxExtent, uint_t BlockSize, class... Funs>
+        struct kernel {
+            tuple<Funs...> m_funs;
+            size_t m_shared_memory_size;
+
+            template <class Grid, class Kernel>
+            Kernel launch_or_fuse(Grid const &grid, Kernel kernel) && {
+                launch_kernel<MaxExtent, i_block_size_t::value, j_block_size_t::value>(grid.i_size(),
+                    grid.j_size(),
+                    blocks_required_z<BlockSize>(grid.k_size()),
+                    make_multi_kernel(std::move(m_funs)),
+                    m_shared_memory_size);
+                return kernel;
+            }
+
+            template <class Grid, class Fun>
+            kernel<MaxExtent, BlockSize, Funs..., Fun> launch_or_fuse(
```

Comment:
As discussed, this may lead to race conditions. In my tests on the V100 and Mi50 actually vertical_advection_dycore pretty much always fails with this fusion enabled.

Reply:
fixed